crates/hulk/src/nao/double_buffered_reader.rs

```diff
@@ -0,0 +1,411 @@
+use std::{
+    io::{self, ErrorKind, Read},
+    mem::size_of,
+    os::unix::{io::AsRawFd, prelude::RawFd},
+};
+
+use super::hula::StateStorage;
+
+pub struct DoubleBufferedReader {
+    buffers: [[u8; size_of::<StateStorage>()]; 2],
+    active_buffer_index: usize,
+    incomplete_offset: usize,
+}
+
+impl Default for DoubleBufferedReader {
+    fn default() -> Self {
+        Self {
+            buffers: [[0; size_of::<StateStorage>()]; 2],
+            active_buffer_index: Default::default(),
+            incomplete_offset: Default::default(),
+        }
+    }
+}
+
+impl DoubleBufferedReader {
+    fn swap_buffers(&mut self) {
+        self.incomplete_offset = 0;
+        self.active_buffer_index = (self.active_buffer_index + 1) % 2;
+    }
+
+    pub fn read(
+        &mut self,
+        reader: &mut (impl AsRawFd + Read),
+        mut poll_reader: impl FnMut(RawFd) -> io::Result<()>,
+    ) -> io::Result<()> {
+        let mut buffers_swapped = false;
```

Comment:
Naming things: This boolean, is used to express whether a full message was received since the last read, correct? Maybe something like `is_buffer_complete`?

Reply:
It is now named `is_at_least_one_buffer_complete`.